gateway/hostname.go

```diff
@@ -0,0 +1,623 @@
+package gateway
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	cid "github.com/ipfs/go-cid"
+	namesys "github.com/ipfs/go-namesys"
+	"github.com/libp2p/go-libp2p/core/peer"
+	dns "github.com/miekg/dns"
+
+	mbase "github.com/multiformats/go-multibase"
+
+	nsopts "github.com/ipfs/interface-go-ipfs-core/options/namesys"
+)
+
+type HostnameConfig struct {
+	// NameSystem is the name system used to resolve TXT records for DNSLink hosts.
+	NameSystem namesys.NameSystem
+
+	// NoDNSLink configures the gateway to _not_ perform DNS TXT record
+	// lookups in response to requests with values in `Host` HTTP header.
+	// This flag can be overridden per FQDN in PublicGateways.
+	NoDNSLink bool
+
+	// PublicGateways configures behavior of known public gateways.
+	// Each key is a fully qualified domain name (FQDN).
+	PublicGateways map[string]*Specification
+}
+
+// Specification is the specification of an IPFS Public Gateway.
+type Specification struct {
+	// Paths is explicit list of path prefixes that should be handled by
+	// this gateway. Example: `["/ipfs", "/ipns", "/api"]`
+	Paths []string
+
+	// UseSubdomains indicates whether or not this gateway uses subdomains
+	// for IPFS resources instead of paths. That is: http://CID.ipfs.GATEWAY/...
+	//
+	// If this flag is set, any /ipns/$id and/or /ipfs/$id paths in Paths
+	// will be permanently redirected to http://$id.[ipns|ipfs].$gateway/.
+	//
+	// We do not support using both paths and subdomains for a single domain
+	// for security reasons (Origin isolation).
+	UseSubdomains bool
+
+	// NoDNSLink configures this gateway to _not_ resolve DNSLink for the FQDN
+	// provided in `Host` HTTP header.
+	NoDNSLink bool
+
+	// InlineDNSLink configures this gateway to always inline DNSLink names
+	// (FQDN) into a single DNS label in order to interop with wildcard TLS certs
+	// and Origin per CID isolation provided by rules like https://publicsuffix.org
+	InlineDNSLink bool
+}
+
+// DefaultPaths are the default path prefixes that are handled by default.
+var DefaultPaths = []string{"/ipfs/", "/ipns/", "/api/", "/p2p/"}
```

Comment:
This may require refactor in Kubo, but we dont want these Kubo-specific thingss in go-libipfs.

```suggestion
// DefaultPaths are the default path prefixes that are handled by default.
var DefaultPaths = []string{"/ipfs/", "/ipns/"}
```